src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -543,28 +549,31 @@ private bool IsPreamble()
         {
             if (!_checkPreamble)
             {
-                return _checkPreamble;
+                return false;
             }
 
-            ReadOnlySpan<byte> preamble = _encoding.Preamble;
+            return IsPreambleWorker(); // move this call out of the hot path
+            bool IsPreambleWorker()
+            {
+                Debug.Assert(_checkPreamble);
+                ReadOnlySpan<byte> preamble = _encoding.Preamble;
 
-            Debug.Assert(_bytePos <= preamble.Length, "_compressPreamble was called with the current bytePos greater than the preamble buffer length.  Are two threads using this StreamReader at the same time?");
-            int len = (_byteLen >= (preamble.Length)) ? (preamble.Length - _bytePos) : (_byteLen - _bytePos);
+                Debug.Assert(_bytePos < preamble.Length, "_compressPreamble was called with the current bytePos greater than the preamble buffer length.  Are two threads using this StreamReader at the same time?");
+                int len = Math.Min(_byteLen, preamble.Length);
 
-            for (int i = 0; i < len; i++, _bytePos++)
-            {
-                if (_byteBuffer[_bytePos] != preamble[_bytePos])
+                for (int i = _bytePos; i < len; i++)
                 {
-                    _bytePos = 0;
-                    _checkPreamble = false;
-                    break;
+                    if (_byteBuffer[i] != preamble[i])
+                    {
+                        _bytePos = 0; // preamble match failed; back up to beginning of buffer
+                        _checkPreamble = false;
+                        return false;
+                    }
                 }
-            }
+                _bytePos = len; // we've matched all bytes up to this point
```

Comment:
I assume it's worth open-coding the loop because it's expected that most preambles are just a few bytes?  If we only do this once per stream on average, would it still be better to pay the few extra ns and just use SequenceEqual in order to simplify the code?